src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameService.cs

```diff
@@ -84,6 +92,38 @@ internal class InlineRenameService : IInlineRenameService
                 _asyncListener);
 
             return new InlineRenameSessionInfo(ActiveSession);
+
+            static InlineRenameSessionInfo IsReadOnlyOrCannotNavigateToSpan(IInlineRenameInfo renameInfo, Document document, CancellationToken cancellationToken)
+            {
+                if (renameInfo is IInlineRenameInfoWithFileRename renameInfoWithFileRename)
+                {
+                    var workspace = document.Project.Solution.Workspace;
+                    var navigationService = workspace.Services.GetService<IDocumentNavigationService>();
+                    foreach (var documentSpan in renameInfoWithFileRename.DefinitionLocations)
+                    {
+                        var sourceText = documentSpan.Document.GetTextSynchronously(cancellationToken);
+                        var textSnapshot = sourceText.FindCorrespondingEditorTextSnapshot();
+
+                        if (textSnapshot != null)
+                        {
+                            var buffer = textSnapshot.TextBuffer;
+                            var originalSpan = documentSpan.SourceSpan.ToSnapshotSpan(textSnapshot).TranslateTo(buffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+
+                            if (buffer.IsReadOnly(originalSpan))
+                            {
+                                return new InlineRenameSessionInfo(EditorFeaturesResources.You_cannot_rename_this_element_because_it_is_contained_in_a_read_only_file);
+                            }
+
+                            if (!navigationService.CanNavigateToSpan(workspace, document.Id, documentSpan.SourceSpan))
```

Comment:
minor optimization - seems like this check could be done before retrieving the sourceText at the top of the loop since it doesn't require any of them.